modules/dnn/src/layers/scatter_layer.cpp

```diff
@@ -0,0 +1,217 @@
+        else if (reduction_name == "mul")
+            reduction = REDUCTION::MUL;
+        else if (reduction_name == "max")
+            reduction = REDUCTION::MAX;
+        else if (reduction_name == "min")
+            reduction = REDUCTION::MIN;
+        else
+            CV_Error(cv::Error::StsBadArg, "Unkown reduction \"" + reduction_name + "\"");
+    }
+
+    virtual bool supportBackend(int backendId) CV_OVERRIDE
+    {
+        return backendId == DNN_BACKEND_OPENCV;
+    }
+
+    virtual bool getMemoryShapes(const std::vector<MatShape> &inputs,
+                                 const int requiredOutputs,
+                                 std::vector<MatShape> &outputs,
+                                 std::vector<MatShape> &internals) const CV_OVERRIDE
+    {
+        CV_CheckEQ(inputs.size(), 3ull, "Scatter: require three inputs.");
+        CV_CheckEQ(inputs[0].size(), inputs[1].size(), "Scatter: input data should have the same ndim with indices.");
+        CV_CheckEQ(inputs[0].size(), inputs[2].size(), "Scatter: input data should have the same ndim with updates.");
+        for (size_t i = 0; i < inputs[0].size(); i++)
+        {
+            CV_CheckGE(inputs[0][i], inputs[1][i], "Scatter: each dim of input data should be greater than (or equal to) indices'.");
+            CV_CheckEQ(inputs[1][i], inputs[2][i], "Scatter: each dim of indices should be equal to updates'.");
+        }
+        outputs.assign(1, inputs[0]);
+        return false;
+    }
+
+    virtual void finalize(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr) CV_OVERRIDE
+    {
+        std::vector<Mat> inputs;
+        inputs_arr.getMatVector(inputs);
+
+        int ndims = inputs[0].dims;
+        // current_indices = std::make_shared<int>(ndims);
+        current_indices.resize(ndims);
+    }
+
+    void forward(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr, OutputArrayOfArrays internals_arr) CV_OVERRIDE
+    {
+        CV_TRACE_FUNCTION();
+        CV_TRACE_ARG_VALUE(name, "name", name.c_str());
+
+        std::vector<Mat> inputs, outputs;
+        inputs_arr.getMatVector(inputs);
+        outputs_arr.getMatVector(outputs);
+
+        const Mat& data = inputs[0];
+        const Mat& indices = inputs[1];
+        const Mat& updates = inputs[2];
+        Mat& out = outputs[0];
+
+        const int ndims = data.dims;
+        const int* shape = data.size.p;
+        const size_t* step = data.step.p;
+
+        const int* ind_shape = indices.size.p;
+        const size_t* ind_step = indices.step.p;
+
+        typeDispatch(outputs[0].type(), ndims, shape, step, data.ptr<char>(), ind_shape, ind_step, indices.ptr<char>(), updates.ptr<char>(), out.ptr<char>());
+    }
+
+    template<typename T, typename Functor>
+    void forward_impl(const Functor& rd, int ndims, const int* shape, const size_t* step, const char* input, const int* ind_shape, const size_t* ind_step, const char* indices, const char* updates, char* out)
```

Comment:
It went much slower (from ~220ms to ~290ms) if passing mat. You could have a look at `forward_impl_mat` in case I did something inefficient.

Reply:
1. It might be a call to `total()` inside each iteration of for loop - let's try to cache it and see if it improves.
2. As for signature, it can be reduced to `void forward_impl(const Functor& rd, const Mat& input, const Mat& indices, const Mat& updates, Mat& out)`
3. What hardware are you testing on? My AMD Ryzen 5 5600H manages to run in 800ms(!), not 200ms.

Reply:
Cached `total()` and it did improve a lot, but still a little slower than the initial version. Also tried to restore current_indices with:
```cpp
current_indices[ndims - 1] += 1;
for (int j = ndims - 1; j > 0; j--)
{
    current_indices[j - 1] += int(current_indices[j] / ind_shape[j]);
    current_indices[j] %= ind_shape[j];
}
```
, and this went way slower (+100ms). Code is updated and you could take a look if you want.

> I still think it's a little bit too much, isn't it?

It depends on the indices shape. The shape of indices in the perf test I set is 8x256x128x100, which is kind of stressful. I checked the models which require ScatterND, and the shapes of indices are like 1x3x80x80x6x5 (1/45 of the stress test) from https://github.com/opencv/opencv/issues/22528, 1x168x2 from https://github.com/opencv/opencv/issues/22541#issuecomment-1252268440. I could not say we can never reach a scale like the stress test, but maybe the current implementation satisfies most use cases, not to mention that a model containing scatter or scatterND is quite rare. We could improve it latter if there is a great demand on the speed of scatter & scatterND.

Reply:
I think +0.5% is negligible. Does caching step_ and indices_step_ improve the performance?
Everything looks good, so please make the same changes to scatterND to be consistent and it will be merged.

Reply:
> Does caching step_ and indices_step_ improve the performance?

Minor improvement like -2ms. I think we dont cache step_ and indices_step_ to save mem and keep the code clean and simple.

> so please make the same changes to scatterND

Sure, will do.